File: python/combinationSum.py
# difficulty: medium
# https://leetcode.com/problems/combination-sum/

# Given a set of candidate numbers (candidates) (without duplicates) and a target number (target), find all unique combinations in candidates where the candidate numbers sums to target.

# The same repeated number may be chosen from candidates unlimited number of times.

# Note:

# All numbers (including target) will be positive integers.
# The solution set must not contain duplicate combinations.
# Example 1:

# Input: candidates = [2,3,6,7], target = 7,
# A solution set is:
# [
#   [7],
#   [2,2,3]
# ]

def combinationSum(self, candidates, target):
    """
    :type candidates: List[int]
    :type target: int
    :rtype: List[List[int]]
    """
    output = []
    def recursive(digits,candidates,value):
        if value == target:
          output.append(digits)
        else:
          for i in range(len(candidates)):
            if value+candidates[i] <= target:
              recursive(digits+[candidates[i]], candidates[i:], value+candidates[i])

    recursive([],candidates,0)
    return output

File: javascript/combinationSum.js
// difficulty: medium
// https://leetcode.com/problems/combination-sum/
//
// Given an array of distinct integers candidates and a target integer target, return a list of all unique combinations of candidates where the chosen numbers sum to target. You may return the combinations in any order.
//
// The same number may be chosen from candidates an unlimited number of times. Two combinations are unique if the frequency of at least one of the chosen numbers is different.
//
// It is guaranteed that the number of unique combinations that sum up to target is less than 150 combinations for the given input.
//
// Example 1:
// Input: candidates = [2,3,6,7], target = 7
// Output: [[2,2,3],[7]]
//
// Example 2:
// Input: candidates = [2,3,5], target = 8
// Output: [[2,2,2,2],[2,3,3],[3,5]]

var combinationSum = function(candidates, target) {
    let ret = []
    function recursive(nums, candidates, value){
        if(value == target) ret.push(nums)
        else{
            for(let i = 0; i < candidates.length; i++){
                if(value + candidates[i] <= target){
                    recursive(nums.concat(candidates[i]), candidates.slice(i), value+candidates[i])
                }
            }
        }
    }
    recursive([],candidates,0)
    return ret
};
